libbeat/publisher/pipeline/output_test.go

```diff
@@ -0,0 +1,181 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package pipeline
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/outputs"
+	"github.com/elastic/beats/v7/libbeat/publisher"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+func TestPublish(t *testing.T) {
+	tests := map[string]struct {
+		client outputs.Client
+	}{
+		"client": {
+			&mockClient{},
+		},
+		"network_client": {
+			&mockNetworkClient{},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			rand.Seed(time.Now().UnixNano())
+
+			wqu := makeWorkQueue()
+			makeClientWorker(nilObserver, wqu, test.client)
+
+			numEvents := atomic.MakeInt(0)
+			var wg sync.WaitGroup
+			for batchIdx := 0; batchIdx <= randIntBetween(25, 200); batchIdx++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					batch := randomBatch(50, 150, wqu)
+
+					numEvents.Add(len(batch.Events()))
+
+					wqu <- batch
+				}()
```

Comment:
we are moving batches into the work queue only. Why do we need a go-routine for this?
In case the queue/worker blocks the test will timeout (default test timeout in go is 10min).

Reply:
Yeah, good point. I was trying to simulate as much as possible how the publisher is being used in reality — e.g. multiple FB inputs concurrently trying to publish batches of events. However, the work queue is an unbuffered channel so trying to concurrently send batches to it is pointless 🤦‍♂. I will remove the goroutine.

Reply:
Fixed in 33af88a42.